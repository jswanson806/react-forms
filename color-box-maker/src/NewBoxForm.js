import React, { useState } from "react";

const NewBoxForm = ({ addBox }) => {
    const INITIAL_STATE = {
        color: '',
        height: '',
        width: ''
    }
    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleSubmit = (e) => {
        e.preventDefault();
        addBox(formData.color, formData.height, formData.width);
        setFormData(INITIAL_STATE);
    }

    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="color">Color: </label>
            <input
                id="color"
                name="color"
                placeholder="color"
                value={formData.color}
                onChange={handleChange} 
            />
            <label htmlFor="height">Height: </label>
            <input
                id="height"
                name="height"
                placeholder="height"
                value={formData.height}
                onChange={handleChange} 
            />
            <label htmlFor="width">Width: </label>
            <input
                id="width"
                name="width"
                placeholder="width"
                value={formData.width}
                onChange={handleChange} 
            />
            <button>Add Box</button>
        </form>
    )
}

export default NewBoxForm;