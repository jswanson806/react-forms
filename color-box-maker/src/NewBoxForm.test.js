import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

it('renders without crashing', function() {
    render(<NewBoxForm />)
})

it('matches the snapshot', function() {
    const { asFragment } = render(<NewBoxForm />)
    expect(asFragment()).toMatchSnapshot();
})

it('handles user input correctly', function() {
    const { getByLabelText } = render(<NewBoxForm />);
    const colorInput = getByLabelText("Color:");
    const heightInput = getByLabelText("Height:");
    const widthInput = getByLabelText("Width:");

    fireEvent.change(colorInput, { target: {value: "red"} })
    fireEvent.change(heightInput, { target: {value: "100px"} })
    fireEvent.change(widthInput, { target: {value: "100px"} })

    expect(colorInput).toHaveValue("red")
    expect(heightInput).toHaveValue("100px")
    expect(widthInput).toHaveValue("100px")
})


it('submits the form data correctly', function() {
    // spy on addBox function
    const addBox = jest.fn();
    const { getByLabelText, getByRole } = render(<NewBoxForm addBox={addBox}/>)
    const colorInput = getByLabelText("Color:");
    const heightInput = getByLabelText("Height:");
    const widthInput = getByLabelText("Width:");
    const btn = getByRole("button", {name: "Add Box"});

    fireEvent.change(colorInput, { target: {value: "red"} })
    fireEvent.change(heightInput, { target: {value: "100px"} })
    fireEvent.change(widthInput, { target: {value: "100px"} })
    fireEvent.click(btn);

    // function call should be made only once
    expect(addBox).toHaveBeenCalledTimes(1);
    // function call should be with correct data
    expect(addBox).toHaveBeenCalledWith("red", "100px","100px")
})