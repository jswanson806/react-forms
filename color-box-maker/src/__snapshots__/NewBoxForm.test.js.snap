// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color: 
    </label>
    <input
      id="color"
      name="color"
      placeholder="color"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      value=""
    />
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      value=""
    />
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
